File: quick sort compare/quick_sort.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

//@brief：耗时统计函数
func timeCost() func() {
	start := time.Now()
	return func() {
		tc := time.Since(start)
		fmt.Printf("time cost = %v\n", tc)
	}
}

func quickSort(arr []int) []int {

	if len(arr) <= 1 {
		return arr
	}

	var left []int
	var right []int
	var middle []int
	middle = append(middle, arr[0])
	for i := 1; i < len(arr); i++ {
		if arr[i] < arr[0] {
			right = append(right, arr[i])
		} else if arr[i] > arr[0] {
			left = append(left, arr[i])
		} else {
			middle = append(middle, arr[i])
		}
	}
	right = quickSort(right)
	left = quickSort(left)
	myarr := append(append(left, middle...), right...)

	return myarr
}

func getarr(n, m int) []int {
	var arr []int
	rand.Seed(9)
	for i := 1; i < n; i++ {
		arr = append(arr, rand.Intn(m)) //生成[0,10000000)的随机数
	}
	return arr
}

func main() {
	defer timeCost()()
	getarr(999999, 1000000)
	//fmt.Println(arr)

	//fmt.Println(quickSort(arr))
	//quickSort(arr)
}

File: quick sort compare/quick_sort.py
import time
import random
import datetime


def quicksort(array,begin,end):
	if begin < end:
		i = begin + 1 
		j = end  
		while 1:
			if i >= j:
				break
			if array[i] > array[begin]:
				array[i], array[j] = array[j], array[i]
				j = j - 1
			else:
				i = i + 1
			
		if array[i] >= array[begin]:
			i = i - 1

		array[begin], array[i] = array[i], array[begin]
		quicksort(array, begin, i)
		quicksort(array, j, end)


def getarr(n, m):
    arr = []
    random.seed(9)
    for i in range(m):
        arr.append(random.randint(1,m))
    return arr

if __name__ == '__main__':
	arr = getarr(99999,100000)
	start=datetime.datetime.now()
	quicksort(arr,0,len(arr)-1)
	end=datetime.datetime.now()
	print(end-start)
